test/e2e/baseline_test.go

```diff
@@ -443,6 +448,132 @@ spec:
 	})
 }
 
+// Test add/remove waypoint at pod granularity.
+func TestAddRemovePodWaypoint(t *testing.T) {
+	framework.NewTest(t).Run(func(t framework.TestContext) {
+		waypoint := "pod-waypoint"
+		newWaypointProxyOrFail(t, t, apps.Namespace, waypoint, constants.WorkloadTraffic)
+
+		t.Cleanup(func() {
+			deleteWaypointProxyOrFail(t, t, apps.Namespace, waypoint)
+		})
+
+		dst := apps.EnrolledToKmesh
+		t.NewSubTest("before").Run(func(t framework.TestContext) {
+			for _, src := range apps.All {
+				if src.Config().IsUncaptured() {
+					continue
+				}
+				for _, dstWl := range dst.WorkloadsOrFail(t) {
+					t.NewSubTestf("from %v", src.Config().Service).Run(func(t framework.TestContext) {
+						c := IsL4()
+						opt := echo.CallOptions{
+							Address: dstWl.Address(),
+							Port:    echo.Port{ServicePort: ports.All().MustForName("http").WorkloadPort},
+							Scheme:  scheme.HTTP,
+							Count:   10,
+							Check:   check.And(check.OK(), c),
+						}
+						src.CallOrFail(t, opt)
+					})
+				}
+
+			}
+		})
+
+		// Configure pods to use waypoint.
+		for _, dstWl := range dst.WorkloadsOrFail(t) {
+			SetWaypoint(t, apps.Namespace.Name(), dstWl.PodName(), waypoint, Workload)
+		}
+
+		// Now should always be L7.
+		t.NewSubTest("after").Run(func(t framework.TestContext) {
+			for _, src := range apps.All {
+				if src.Config().IsUncaptured() {
+					continue
+				}
+				for _, dstWl := range dst.WorkloadsOrFail(t) {
+					t.NewSubTestf("from %v", src.Config().Service).Run(func(t framework.TestContext) {
+						c := IsL4()
+						opt := echo.CallOptions{
+							Address: dstWl.Address(),
+							Port:    echo.Port{ServicePort: ports.All().MustForName("http").WorkloadPort},
+							Scheme:  scheme.HTTP,
+							Count:   10,
+							Check:   check.And(check.OK(), c),
+						}
+						src.CallOrFail(t, opt)
+					})
+				}
+
+			}
+		})
+	})
+}
+
+// Test add/remove waypoint at ns or service granularity.
+func TestRemoveAddNsOrServiceWaypoint(t *testing.T) {
+	for _, granularity := range []Granularity{Service /*,Namespace*/} {
```

Comment:
@hzxuzhonghu maybe this PR could be merged first, when https://github.com/kmesh-net/kmesh/issues/628 get fixed, then we uncomment `Namespace`